### PR TITLE
libxdp: make shared xsk creation network namespace aware

### PR DESCRIPTION
**Pull request by Dhiver**
Creating AF_XDP sockets with the XDP_SHARED_UMEM flag set fails when two devices
ifindex (and the queue_id) are the same. This can happen with devices in
different network namespaces.

In the shared UMEM mode, each Xsk is expected to have a dedicated fill and
completion queue. Context data about those queues is stored in a linked-list
within the UMEM object.
The comparison performed to pick the right context for an Xsk prior to calling
bind() only takes into account the ifindex and queue_id of the network device.
However, the ifindex is not a "system-wide" unique identifier, it is unique only
within a network namespace. This results in attributing the same xsk_ctx to
different Xsks.

To fix this, we retrieve the netns_cookie associated with the socket at its
creation and add it in the comparison operation. This value "provides a global
identifier that can be assumed unique" ([0]).
We use the netns cookie instead of the netns inode number because inode numbers
can be reused.

  [0]: https://lore.kernel.org/bpf/c47d2346982693a9cf9da0e12690453aded4c788.1585323121.git.daniel@iogearbox.net/

### CODE REVIEW

**Comment by tohojo** [comment]
Note that the socket option to retrieve the netns cookie was only added by commit `e8b9eab99232 ("net: retrieve netns cookie via getsocketopt")` in kernel 5.14, so failing when that is not available breaks things on older kernels. Probably not a good idea to do that unconditionally...

**Comment by tohojo** [comment]
How do you end up creating multiple sockets in different namespaces using the same umem object anyway?

File: lib/libxdp/xsk.c [hunk 1]
```diff
@@ -898,14 +902,20 @@ int xsk_socket__create_shared(struct xsk_socket **xsk_ptr,
 		tx_setup_done = umem->tx_ring_setup_done;
 	}
 
-	ctx = xsk_get_ctx(umem, ifindex, queue_id);
+	optlen = sizeof(netns_cookie);
+	if (getsockopt(xsk->fd, SOL_SOCKET, SO_NETNS_COOKIE, &netns_cookie, &optlen) != 0) {
+		err = -errno;
+		goto out_socket;
+	}
```
Comment (alexforster) [inline lib/libxdp/xsk.c:909]: Given [this comment](https://github.com/xdp-project/xdp-tools/pull/205#issuecomment-1179375437), maybe we could conditionally compile this `getsockopt` call and instead set `netns_cookie` to 0 on kernels older than 5.14.

Question for maintainers: is use of the `KERNEL_VERSION` macro acceptable here?

**Comment by alexforster** [comment]
> How do you end up creating multiple sockets in different namespaces using the same umem object anyway?

We're attaching AF_XDP to both sides of a veth pair and sharing a Umem between them. We use this setup to inspect/manipulate traffic going into and out of containers.

File: lib/libxdp/xsk.c (hunk 1 above)
Comment (tohojo) [reply lib/libxdp/xsk.c:909]: Nope, can't use `KERNEL_VERSION`, features may be backported. You can just detect the return value at runtime, though, and fall back to setting a fixed value. Something like this:

https://github.com/xdp-project/bpf-examples/commit/e92fd14fafb97603546e82814450d92396cd3636

**Comment by tohojo** [comment]
> > How do you end up creating multiple sockets in different namespaces using the same umem object anyway?
> 
> We're attaching AF_XDP to both sides of a veth pair and sharing a Umem between them. We use this setup to inspect/manipulate traffic going into and out of containers.

Hmm, interesting. What do you gain by sharing the umem instead of having completely separate attachments on each side?

File: lib/libxdp/xsk.c (hunk 1 above)
Comment (Dhiver) [reply lib/libxdp/xsk.c:919]: Thank you for the example. I've updated the PR based on this feedback (see commit 2c9573f88deb1587630285b3567f82894aa2a955).

**Comment by Dhiver** [comment]
> > > How do you end up creating multiple sockets in different namespaces using the same umem object anyway?
> > 
> > 
> > We're attaching AF_XDP to both sides of a veth pair and sharing a Umem between them. We use this setup to inspect/manipulate traffic going into and out of containers.
> 
> Hmm, interesting. What do you gain by sharing the umem instead of having completely separate attachments on each side?

This allows for zero-copy between the two veth. UMEM frames don't need to be copied from one UMEM to the other.

File: lib/libxdp/xsk.c (hunk 1 above)
Comment (tohojo) [reply lib/libxdp/xsk.c:919]: Alright, cool! Please squash the fixup into the original commit, though... :)

**Comment by tohojo** [comment]
> This allows for zero-copy between the two veth. UMEM frames don't need to be copied from one UMEM to the other.

Right, okay, but, erm, I must be missing something because by that description it seems like you have the same process open AF_XDP sockets inside and outside the container and zero-copy frames to itself through the veth pair? Seems a bit of a roundabout way of doing shared memory? :)

File: lib/libxdp/xsk.c (hunk 1 above)
Comment (Dhiver) [reply lib/libxdp/xsk.c:919]: squashed

**Comment by Dhiver** [comment]
> > This allows for zero-copy between the two veth. UMEM frames don't need to be copied from one UMEM to the other.
> 
> Right, okay, but, erm, I must be missing something because by that description it seems like you have the same process open AF_XDP sockets inside and outside the container and zero-copy frames to itself through the veth pair? Seems a bit of a roundabout way of doing shared memory? :)

We are moving frame **descriptors** between the TX and RX queues of the AF_XDP sockets that we created on both interfaces of the veth pair (which are in different netns). That way we can forward, send back or drop packets depending on which Xsk RX/TX queue we produce the descriptor. We can also edit the frame content from our userspace application. I hope that this makes more sense.

**Comment by tohojo** [comment]
> > > This allows for zero-copy between the two veth. UMEM frames don't need to be copied from one UMEM to the other.
> > 
> > 
> > Right, okay, but, erm, I must be missing something because by that description it seems like you have the same process open AF_XDP sockets inside and outside the container and zero-copy frames to itself through the veth pair? Seems a bit of a roundabout way of doing shared memory? :)
> 
> We are moving frame **descriptors** between the TX and RX queues of the AF_XDP sockets that we created on both interfaces of the veth pair (which are in different netns). That way we can forward, send back or drop packets depending on which Xsk RX/TX queue we produce the descriptor. We can also edit the frame content from our userspace application. I hope that this makes more sense.

Ah, and the frames themselves are coming in on a physical NIC in the host namespace? That makes sense, then; interesting!

Anyway, I don't have any objections to merging this, but let's see what @magnus-karlsson says :)

**Comment by magnus-karlsson** [comment]
I am all for this fix. Thanks!

**Review by tohojo — APPROVED** [review]

**Comment by Dhiver** [comment]
For more context and visual representation @tohojo : https://blog.cloudflare.com/a-story-about-af-xdp-network-namespaces-and-a-cookie/ ;)

**Pull request merged**